File: sicp_1_30.swift
#!/usr/bin/env swift
// sicp exercise 1.30

func sum(low: Int, high: Int, term: (Int) -> (Int), next: (Int) -> (Int)) -> Int {
  return sumIter(low, 0, high, term, next)
}

func sumIter(a: Int, result: Int, b: Int, term: (Int) -> (Int), next: (Int) -> (Int)) -> Int {
  return a > b ? result : sumIter(next(a), (term(a)+result), b, term, next)
}

func inc(n: Int) -> Int {
  return n + 1
}

func cube(n: Int) -> Int {
  return (n * n * n)
}

func sumCubes(a: Int, b: Int) -> Int {
  return sum(a, b, cube, inc)
}

println(sumCubes(1,10))

File: sicp_1_30.py
#!/usr/bin/env python
# sicp 1.30

def sum(low, high, term, next):
  return sum_iter(low, 0, high, term, next)

def sum_iter(a, result, b, term, next):
  if a > b:
    return result
  else:
    return sum_iter(next(a), (term(a)+result), b, term, next)

def inc(n):
  return n + 1

def cube(n):
  return n * n * n

def sum_cubes(a,b):
  return sum(a,b,cube,inc)

print(sum_cubes(1,10))
